{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#To convert this to a Python script, use:\n",
    "#jupyter nbconvert --to python Mikado.ipynb\n",
    "\n",
    "#To run on a SLURM scheduler or other shared computing system, you may need to load modules.\n",
    "#For Savio, utilize the script Mikado.sh\n",
    "\n",
    "#To upload this to SLURM, along with other necessary files, use\n",
    "#scp Mikado.py Mikado.sh dasun@dtn.brc.berkeley.edu:/global/scratch/dasun/Transcriptome_assessment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading libraries\n",
      "loaded libraries\n"
     ]
    }
   ],
   "source": [
    "#Configuration of Python script\n",
    "!printf 'loading libraries\\n'\n",
    "import subprocess\n",
    "import os\n",
    "import os.path\n",
    "import fnmatch\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import datetime\n",
    "import matplotlib.pyplot as plt\n",
    "from io import StringIO\n",
    "import sys\n",
    "!printf 'loaded libraries\\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "working on $transcriptome and outputting to $output_filezsh:1: command not found: gmap\n",
      "working on $transcriptome and outputting to $output_filezsh:1: command not found: gmap\n",
      "working on $transcriptome and outputting to $output_filezsh:1: command not found: gmap\n",
      "working on $transcriptome and outputting to $output_filezsh:1: command not found: gmap\n",
      "working on $transcriptome and outputting to $output_filezsh:1: command not found: gmap\n"
     ]
    }
   ],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to run this section\n",
    "run_gmap_build = False\n",
    "run_gmap_align = False\n",
    "run_gmap_prepare = False\n",
    "run_samtools_sort = False\n",
    "run_picardtools_valsam = False\n",
    "run_picardtools_clean = False\n",
    "run_portcullis = False\n",
    "run_samtools_clean = False\n",
    "run_portcullis_insilico = False\n",
    "\n",
    "#Write a base pattern for your raw fastq file names\n",
    "#This pattern should occur in all of your file names\n",
    "genome_file = 'phaw_5.0.fa'\n",
    "alignment_bam = 'Sall_RNA_sorted.aligned.bam'\n",
    "transcriptome_list = [\n",
    "    'Phaw_DAS_Trinity-GG_hisat2.fasta'\n",
    "]\n",
    "old_transcriptome_list = [\n",
    "    'HB_Transcriptome_Trinity-GG.fasta',\n",
    "    'HB_Transcriptome_Trinity.fasta',\n",
    "    'ParhyaleTranscriptome.ReAnnotated.reformatted4.fasta',\n",
    "    'Phaw_DAS_Trinity.fasta',\n",
    "    'Trinity_S13.fasta',\n",
    "    'Trinity_S19.fasta',\n",
    "    'Trinity_S21.fasta',\n",
    "    'Trinity_S23.fasta',\n",
    "    'PhRef5and3RACE.fa',\n",
    "    'Phaw.Kao.annotations.transcript.fa'\n",
    "]\n",
    "gff_list = [\n",
    "    'Phaw_DAS_Trinity-GG_hisat2.gff'\n",
    "]\n",
    "old_gff_list = [\n",
    "    'HB_Transcriptome_Trinity-GG.gff',\n",
    "    'HB_Transcriptome_Trinity.gff',\n",
    "    'ParhyaleTranscriptome.ReAnnotated.reformatted4.gff',\n",
    "    'Phaw.Kao.annotations.transcript.gff',\n",
    "    'Phaw_DAS_Trinity.gff',\n",
    "    'Trinity_S13.gff',\n",
    "    'Trinity_S19.gff',\n",
    "    'Trinity_S21.gff',\n",
    "    'Trinity_S23.gff'\n",
    "]\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "genome_name = genome_file.replace('.fa', '')\n",
    "genome_dir = genome_name + '_dir'\n",
    "\n",
    "#checks if you want to trim your reads\n",
    "if run_gmap_build == True:\n",
    "    !gmap_build -s=none -D $genome_dir -d $genome_name $genome_file\n",
    "\n",
    "if run_gmap_align == True:\n",
    "    for transcriptome in transcriptome_list:\n",
    "        output_file = transcriptome.replace('.fasta', '.gff')\n",
    "        !printf 'working on '$transcriptome' and outputting to '$output_file'\\n'\n",
    "        !/clusterfs/vector/home/groups/software/sl-7.x86_64/modules/gmap-gsnap/2017-10-30/bin/gmap -D $genome_dir --max-intronlength-ends=300000 --format=gff3_gene --nthreads=20 -d $genome_name $transcriptome > $output_file\n",
    "        \n",
    "if run_gmap_prepare == True:\n",
    "    for gff in gff_list:\n",
    "        gff_sorted = gff.replace('.gff', '.sorted.gff')\n",
    "        !printf 'igvtools sort on '$gff' and outputting to '$gff_sorted'\\n'\n",
    "        !igvtools sort $gff $gff_sorted\n",
    "        !printf 'igvtools index on '$gff_sorted'\\n'\n",
    "        !igvtools index $gff_sorted\n",
    "\n",
    "if run_samtools_sort == True:\n",
    "    !samtools view -F 0x04 -b Sall_RNA_sorted.bam > Sall_RNA_sorted.aligned.bam\n",
    "\n",
    "if run_picardtools_valsam == True:\n",
    "    !java -jar /clusterfs/vector/home/groups/software/sl-7.x86_64/modules/picard/2.9.0/lib/picard.jar ValidateSamFile I=Sall_RNA_sorted.aligned.bam MODE=SUMMARY\n",
    "\n",
    "if run_picardtools_clean == True:\n",
    "    !java -jar /clusterfs/vector/home/groups/software/sl-7.x86_64/modules/picard/2.9.0/lib/picard.jar CleanSam I=Sall_RNA_sorted.aligned.bam O=Sall_RNA_sorted.cleaned.bam\n",
    "\n",
    "if run_portcullis == True:\n",
    "    cleaned_bam = 'Sall_RNA_sorted.bam'\n",
    "    portcullis_out = alignment_bam.replace('.bam', '.portcullis.bed')\n",
    "    now = datetime.now()\n",
    "    dt_string = now.strftime(\"%Y-%m-%d_%H.%M.%S\")\n",
    "    portcullisdebug = df_string + '_portcullis_debug.txt'\n",
    "    !printf 'running portcullis on '$cleaned_bam' and outputting to portcullis_out. stderr and stdout to '$portcullisdebug'/\\n'\n",
    "    !portcullis full --verbose -t 32 $genome_file $cleaned_bam 2>&1 > $portcullisdebug\n",
    "\n",
    "if run_samtools_clean == True:\n",
    "    !samtools view -b -F 4 Phaw_RNA_insilico_q10.bam > Phaw_RNA_insilico_q10.mapped.bam\n",
    "\n",
    "if run_portcullis_insilico == True:\n",
    "    cleaned_bam = 'Phaw_RNA_insilico_q10.mapped.bam'\n",
    "    portcullis_out = alignment_bam.replace('.bam', '.portcullis.bed')\n",
    "    now = datetime.now()\n",
    "    dt_string = now.strftime(\"%Y-%m-%d_%H.%M.%S\")\n",
    "    portcullisdebug = dt_string + '_portcullis_insilico_debug.txt'\n",
    "    !printf 'running portcullis on '$cleaned_bam' and outputting to portcullis_out. stderr and stdout to '$portcullisdebug'/\\n'\n",
    "    !portcullis full --force --copy --verbose -t 32 $genome_file $cleaned_bam 2>&1 | tee $portcullisdebug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to run this section\n",
    "run_hisat2_build = False\n",
    "run_hisat2 = False\n",
    "run_clean_hisat = False\n",
    "run_portcullis_hisat2 = False\n",
    "rerun_portcullis_hisat2 = False\n",
    "\n",
    "genome_file = 'phaw_5.0.fa'\n",
    "index_base = genome_file.replace('.fa', '')\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "#checks if you want to trim your reads\n",
    "if run_hisat2_build == True:\n",
    "    !hisat2-build -p 32 $genome_file $index_base\n",
    "\n",
    "if run_hisat2 == True:\n",
    "    r1 = 'S13A1_S84_R1_001_rd1_1_all_1.fq.gz_ext_all_reads.normalized_K25_C50_pctSD10000.fq'\n",
    "    r2 = 'S13A1_S84_R2_001_rd1_2_all_2.fq.gz_ext_all_reads.normalized_K25_C50_pctSD10000.fq'\n",
    "    Sfile = 'Phaw_RNA_hisat2.sam'\n",
    "    !hisat2 -x $index_base -1 $r1 -2 $r2 -S 'Phaw_RNA_hisat2.sam' --novel-splicesite-outfile 'Phaw_RNA_hisat2_novelsplice.tsv'\n",
    "\n",
    "if run_clean_hisat == True:\n",
    "    start_sam = 'Phaw_RNA_hisat2.sam'\n",
    "    cleaned_bam = start_sam.replace('.sam', '.cleaned.bam')\n",
    "    !samtools view -b -F 4 $start_sam > $cleaned_bam\n",
    "    \n",
    "if run_portcullis_hisat2 == True:\n",
    "    cleaned_bam = 'Phaw_RNA_hisat2.cleaned.bam'\n",
    "    now = datetime.now()\n",
    "    dt_string = now.strftime(\"%Y-%m-%d_%H.%M.%S\")\n",
    "    portcullisdebug = dt_string + '_portcullis_insilico_debug.txt'\n",
    "    !printf 'running portcullis on '$cleaned_bam' and outputting to portcullis_out. stderr and stdout to '$portcullisdebug'/\\n'\n",
    "    !portcullis full --force --copy --verbose -t 32 $genome_file $cleaned_bam 2>&1 | tee $portcullisdebug\n",
    "\n",
    "if rerun_portcullis_hisat2 == True:\n",
    "    !cat portcullis_out/2-junc/portcullis_all.junctions.tab |awk '!(($11==\"?\" && $14==\"NA\") || ($11==\"?\" && $13==\"NA\"))' > portcullis_out/2-junc/portcullis_all.junctions.cleaned.tab\n",
    "    !portcullis filter --verbose portcullis_out/1-prep portcullis_out/2-junc/portcullis_all.junctions.cleaned.tab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zsh:1: command not found: mikado\n"
     ]
    }
   ],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to run this section\n",
    "run_mikado_configure = False\n",
    "\n",
    "mikado_list = '20210630_list.txt'\n",
    "genome_file = '~/Labwork/Bioinformatics/GenomeSequences/phaw_5.0.fa'\n",
    "junction_bed = 'XXX'\n",
    "uniprotloc = '~/Labwork/Bioinformatics/Transcripts/uniprot_sprot.fasta'\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "#checks if you want to trim your reads\n",
    "if run_mikado_configure == True:\n",
    "    !source activate mikado\n",
    "    !mikado configure --list $mikado_list --reference $genome_file --mode permissive --scoring mammalian.yaml --copy-scoring mammalian.yaml -bt $uniprotloc phaw_configuration.yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to run this section\n",
    "run_blastx_prepare = False\n",
    "run_mikado_blastx = False\n",
    "run_transdecoder = False\n",
    "\n",
    "mikado_prepare_fasta = 'mikado_prepared.fasta'\n",
    "uniprot = 'uniprot_sprot.fasta'\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "#checks if you want to trim your reads\n",
    "if run_blastx_prepare == True:\n",
    "    !makeblastdb -in $uniprot -dbtype prot -parse_seqids > blast_prepare.log\n",
    "\n",
    "if run_mikado_blastx == True:\n",
    "    mikado_blast = mikado_prepare_fasta.replace('fasta', 'blast.tsv')\n",
    "    !blastx -max_target_seqs 5 -outfmt \"6 qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore ppos btop\" -num_threads 10 -query $mikado_prepare_fasta -db $uniprot -out $mikado_blast\n",
    "\n",
    "if run_transdecoder == True:\n",
    "    !TransDecoder.LongOrfs -t $mikado_prepare_fasta && TransDecoder.Predict -t $mikado_prepare_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'approximately 0.5795994993742177 genes per second'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'approximately 3924486 genes remain'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "78.36841080040308"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'expected finish at 2021-09-29 19:38:30.693155'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD4CAYAAADsKpHdAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAoeElEQVR4nO3dd3xUZdr/8c8lJfTQAgYChhI6hBIQWXtFXEVUEF3FddlF/emzq2B3rY9l2RULdlZd27oaiiAusBbsIghuOi30ECBAICSU1Pv3x5z4jDGEAElmJvm+X695ceY+5565JiT5zim5xpxziIiInBDoAkREJDgoEEREBFAgiIiIR4EgIiKAAkFERDz1A13AsWrbtq2Ljo4OdBkiIiFlxYoVu5xzEeWtC9lAiI6OZvny5YEuQ0QkpJjZpsOt0yEjEREBFAgiIuJRIIiICKBAEBERjwJBREQABYKIiHgUCCIiAigQRERCxqHCYp5YuJKtew9Wy+MfMRDMrJGZLTOzRDNLNbOHvfGHzGyrmSV4t1F+c+4xs3QzW21mF/iNDzGzZG/ddDMzbzzMzN73xpeaWXQ1vFYRkZCVsGUvo6Z/zStfrmfxqqxqeY7K/KVyPnC2cy7PzBoA35jZQm/d0865J/03NrM+wHigL9AB+NTMejjnioGXgEnA98ACYCSwEJgI7HHOdTez8cBU4Mrjf3kiIqGtoKiE5xav5cUv1tG+eRjvTDyZU2PaVstzHTEQnO8j1fK8uw28W0UfszYaeM85lw9sMLN0YJiZbQRaOOeWAJjZW8Cl+AJhNPCQN38W8LyZmdPHuYlIHbZq+z4mv59I2rZ9XD44igcv6UOLRg2q7fkqdQ7BzOqZWQKQBXzinFvqrbrFzJLM7HUza+WNdQS2+E3P8MY6estlx382xzlXBOQAbY7+5YiIhL7iEsdLX6zjkue+JSv3EDOuHcK0cbHVGgZQyUBwzhU75wYCUfje7ffDd/inGzAQ2AZM8za38h6igvGK5vyMmU0ys+Vmtnznzp2VKV1EJKRs3LWfca8sYeqiVZzbpx0f33YG5/c9sUae+6iuMnLO7QW+AEY653Z4QVEC/B0Y5m2WAXTymxYFZHrjUeWM/2yOmdUHwoHscp5/hnMuzjkXFxFRbvdWEZGQVFLieGvJRi589mvW7sjl2fEDeeHqwbRu2rDGaqjMVUYRZtbSW24MnAusMrNIv83GACne8ofAeO/KoS5ADLDMObcNyDWz4d7VRROAeX5zrvOWrwAW6/yBiNQVmXsPMuH1ZTwwL5VhXVrz8W1nMHpgR7wLMWtMZa4yigTeNLN6+AIk3jn3kZm9bWYD8R3a2QjcAOCcSzWzeCANKAJu9q4wArgJeANojO9kcunVSq8Bb3snoLPxXaUkIlKrOeeY/eNWHv4wlWLneHxMf64a1qnGg6CUheob8bi4OKcPyBGRULUzN597P0jmk7QdDItuzZNjY+ncpkm1P6+ZrXDOxZW3LmQ/MU1EJFQtTN7GfXNTyMsv4s8X9eb6X3Wh3gmB2Svwp0AQEakhOQcKefDDFOYmZNK/YzhPjYslpn3zQJf1EwWCiEgN+HLNTu6clcjuvAJuO7cH/++sbjSoF1zt5BQIIiLVaH9+EY8tWMm7SzcT064Zr04YSv+o8ECXVS4FgohINVm2IZspMxPI2HOQG07vym3n9aBRg3qBLuuwFAgiIlXsUGEx0z5ezavfbKBTqybE33AKQ6NbB7qsI1IgiIhUoaSMvUyOTyQ9K49rhnfmngt70zQsNH7VhkaVIiJBrrC4hOcXp/P85+lENAvjrd8N4/QeodViR4EgInKc1u7IZXJ8Islbc7hsUEcevLgv4U2qtzNpdVAgiIgco+ISx2vfrOfJj9fQPKw+L18zmJH9Io88MUgpEEREjsGm3fu5fWYiP2zcw/l92vP4Zf1p2yws0GUdFwWCiMhRcM7xz6WbeXzBSuqdYDw1LpYxg2q+M2l1UCCIiFTS9pxD3Dk7ia/W7OS0mLZMvXwAHVo2DnRZVUaBICJyBM455iZs5cF5qRQWO/730n5cc3LnWrFX4E+BICJSgd15+dz3QQqLUrcTd1IrnhwbS3TbpoEuq1ooEEREDuM/qdu5d04yuYeKuOfCXvz+tK5B0aa6uigQRETKyDlYyMPzU5nz41b6dmjBu38YSM8Tg6dNdXVRIIiI+Pl67U7unJVEVm4+fzy7O7ecHUPD+sHVprq6KBBERIADBUU8sWAVb3+/iW4RTZlz0whiO7UMdFk1SoEgInXeik3ZTIlPZFP2AX5/ahduv6BnULepri4KBBGpsw4VFvP0p2v4+1fr6dCyMf/6w3CGd20T6LICRoEgInVSytYcJscnsGZHHlcN68x9F/WmWYi0qa4udfvVi0idU1RcwotfrGP6Z2tp3bQh/7h+KGf1bBfosoKCAkFE6oz0rFymxCeSmJHDJbEdeGR0X1o2aRjosoKGAkFEar2SEsfr327gb/9ZTZOG9Xjh6sFcNCB021RXFwWCiNRqW7IPcPvMRJZuyObc3u14/LL+tGveKNBlBSUFgojUSs453vthC49+lMYJZvztigFcMSSq1jWkq0pH/PM7M2tkZsvMLNHMUs3sYW+8tZl9YmZrvX9b+c25x8zSzWy1mV3gNz7EzJK9ddPN+58xszAze98bX2pm0dXwWkWkjtix7xDXv/ED98xJJrZTSxbddjpj4zopDI6gMn+PnQ+c7ZyLBQYCI81sOHA38JlzLgb4zLuPmfUBxgN9gZHAi2ZW+hceLwGTgBjvNtIbnwjscc51B54Gph7/SxORusY5x7yErZz/9Fd8v343D1/Sl3cmnkzHWvSZBdXpiIHgfPK8uw28mwNGA296428Cl3rLo4H3nHP5zrkNQDowzMwigRbOuSXOOQe8VWZO6WPNAs4xRbmIHIXs/QXc8u5/+dN7CXSNaMqCP57GdSOiOaEWdyetapU6h+C9w18BdAdecM4tNbP2zrltAM65bWZWeiFvR+B7v+kZ3liht1x2vHTOFu+xiswsB2gD7CpTxyR8exh07ty5sq9RRGq5T9N2cPecZHIOFnDnyJ5MOq0r9evVjYZ0ValSgeCcKwYGmllL4AMz61fB5uXFsatgvKI5ZeuYAcwAiIuL+8V6Ealbcg8V8sj8NGauyKB3ZAvenjiM3pEtAl1WyDqqq4ycc3vN7At8x/53mFmkt3cQCWR5m2UAnfymRQGZ3nhUOeP+czLMrD4QDmQf5WsRkTrku/Rd3DEriW05B7nlrO788Zy606a6ulTmKqMIb88AM2sMnAusAj4ErvM2uw6Y5y1/CIz3rhzqgu/k8TLv8FKumQ33zg9MKDOn9LGuABZ75xlERH7mYEExD32YytWvLiWs/gnMvmkEt1/QU2FQBSqzhxAJvOmdRzgBiHfOfWRmS4B4M5sIbAbGAjjnUs0sHkgDioCbvUNOADcBbwCNgYXeDeA14G0zS8e3ZzC+Kl6ciNQuP27ew5T4RDbs2s9vR0Rz18heNG5Y99pUVxcL1TficXFxbvny5YEuQ0RqQH5RMc9+upaXv1xHZHhj/jZ2ACO6tQ10WSHJzFY45+LKW6e/VBaRoJaWuY/J8Qms2p7LlXGd+POve9O8UYNAl1UrKRBEJCgVFZfwylfreebTNbRs0pDXrovjnN7tA11WraZAEJGgs25nHlPiE0nYspeLBkTy6Oh+tGqqNtXVTYEgIkGjpMTx5pKNTF20ikYN6vHcVYO4OLZDoMuqMxQIIhIUMvYc4I6ZSSxZv5uzekYw9fIBtGuhNtU1SYEgIgHlnGPm8gwe+SgN5xx/uaw/Vw5VZ9JAUCCISMBk5R7intnJfLYqi5O7tObJsbF0at0k0GXVWQoEEQmIj5Iy+fPcFA4WFPPAr/vwW3UmDTgFgojUqL0HCrh/XirzEzOJ7dSSaWNj6d6uWaDLEhQIIlKDPl+VxZ2zk9izv4Dbz+/BjWd0U5vqIKJAEJFql5dfxKMfpfHeD1vo2b45b1w/lL4dwgNdlpShQBCRarVk3W7umJVI5t6D3HhGN247L4aw+mpIF4wUCCJSLQ4VFvPXRat5/dsNRLdpwswbT2HISa0DXZZUQIEgIlUuccteJscnsG7nfq475STuurAXTRrq102w0/+QiFSZgqISnlu8lhe/WEf75mG8M/FkTo1Rm+pQoUAQkSqxavs+Jr+fSNq2fVw+OIoHL+lDC7WpDikKBBE5LsUljhlfrefpT9bQonF9Zlw7hPP7nhjosuQYKBBE5Jht3LWfKTMTWbFpDxf2O5FHL+1Hm2ZhgS5LjpECQUSOWkmJ452lm3hiwSoa1DOeuXIgowd2UEO6EKdAEJGjkrn3IHfOSuKb9F2c0cPXpvrEcLWprg0UCCJSKc45Zv+4lYc/TKXYOR4f05+rhqlNdW2iQBCRI9qZm8+9HyTzSdoOhkX72lR3bqM21bWNAkFEKrQweRv3zU0hL7+I+0b15nendqGe2lTXSgoEESlXzoFCHvwwhbkJmfTvGM5T42KJad880GVJNVIgiMgvfLlmJ3fOSmR3XgG3nhvDzWd1p4HaVNd6CgQR+cn+/CIeW7CSd5duJqZdM16dMJT+UWpTXVcoEEQEgGUbsrl9ZiJb9hzghtO7ctt5PWjUQG2q65Ij7gOaWScz+9zMVppZqpn9yRt/yMy2mlmCdxvlN+ceM0s3s9VmdoHf+BAzS/bWTTfvejUzCzOz973xpWYWXQ2vVUTKcaiwmMf+ncaVM5YAEH/DKdwzqrfCoA6qzB5CETDFOfejmTUHVpjZJ966p51zT/pvbGZ9gPFAX6AD8KmZ9XDOFQMvAZOA74EFwEhgITAR2OOc625m44GpwJXH//JEpCJJGXuZHJ9IelYevzm5M/eO6k3TMB04qKuO+D/vnNsGbPOWc81sJdCxgimjgfecc/nABjNLB4aZ2UaghXNuCYCZvQVcii8QRgMPefNnAc+bmTnn3LG8KBGpWGFxCc8vTuf5z9OJaBbGm78bxhk9IgJdlgTYUV024B3KGQQs9YZuMbMkM3vdzFp5Yx2BLX7TMryxjt5y2fGfzXHOFQE5QJtynn+SmS03s+U7d+48mtJFxLN2Ry6Xvfgdz362lktiO/CfW09XGAhwFIFgZs2A2cCtzrl9+A7/dAMG4tuDmFa6aTnTXQXjFc35+YBzM5xzcc65uIgIfQOLHA1fm+p1XPTcN2TuPcjL1wzm6SsHEt5En1kgPpU6WGhmDfCFwT+dc3MAnHM7/Nb/HfjIu5sBdPKbHgVkeuNR5Yz7z8kws/pAOJB9tC9GRMq3afd+7piZxLKN2Zzfpz2PX9aftmpTLWVU5iojA14DVjrnnvIbj/TbbAyQ4i1/CIz3rhzqAsQAy7xzEblmNtx7zAnAPL8513nLVwCLdf5A5Pg553jn+01c+OzXrNy+j6fGxfLKtUMUBlKuyuwh/Aq4Fkg2swRv7F7gKjMbiO/QzkbgBgDnXKqZxQNp+K5Qutm7wgjgJuANoDG+k8kLvfHXgLe9E9DZ+K5SEpHjsD3nEHfOTuKrNTs5LaYtUy8fQIeWjQNdlgQxC9U34nFxcW758uWBLkMk6DjnmJuwlQfnpVJY7Lh3VC+uGX6S2lQLAGa2wjkXV946XXAsUovszsvnvg9SWJS6nSEntWLa2Fii2zYNdFkSIhQIIrXEx6nbufeDZPYdLOKeC3vx+9O6qk21HBUFgkiIyzlYyMPzU5nz41b6dmjBP38/kJ4nqk21HD0FgkgI+3rtTu6clURWbj5/PLs7t5wdQ8P6alMtx0aBIBKCDhQU8cSCVbz9/Sa6RTRlzk0jiO3UMtBlSYhTIIiEmBWbspkSn8im7ANMPLULd1zQU51JpUooEERCRH5RMU99soa/f7WeDi0b868/DGd411+0/BI5ZgoEkRCQsjWHKfGJrN6Ry1XDOnHfRX1opjbVUsX0HSUSxIqKS3jxi3VM/2wtrZs25B/XD+Wsnu0CXZbUUgoEkSCVnpXLlPhEEjNyuCS2A4+M7kvLJg0DXZbUYgoEkSBTUuJ4/dsN/O0/q2nSsB4vXD2YiwZEHnmiyHFSIIgEkS3ZB7h9ZiJLN2Rzbu92PH5Zf9o1bxTosqSOUCCIBAHnHO/9sIVHP0rDzPjrFQMYOyRKDemkRikQRAJsx75D3D07ic9X72REtzb89YoBRLVqEuiypA5SIIgEiHOO+UnbuH9uCvlFxTx8SV+uHX4SJ6ghnQSIAkEkALL3F3D/3BT+nbyNQZ1bMm1sLF0jmgW6LKnjFAgiNezTtB3cPSeZnIMF3HFBT244vSv166khnQSeAkGkhuQeKuSR+WnMXJFBrxOb8/bEYfSObBHoskR+okAQqQHfpe/ijllJbMs5yM1ndeNP5/RQm2oJOgoEkWp0sKCYqYtW8cZ3G+natimzbxrBoM6tAl2WSLkUCCLV5MfNe7g9PpH1u/bz2xHR3DWyF40bqk21BC8FgkgVyy8q5tlP1/Lyl+uIDG/Mu384mRHd2ga6LJEjUiCIVKG0zH1Mjk9g1fZcxsVFcf+v+9C8UYNAlyVSKQoEkSpQVFzCK1+t55lP1xDeuCGvTojj3D7tA12WyFFRIIgcp/U785gyM5H/bt7LRQMieXR0P1o1VZtqCT0KBJFjVFLieHPJRqYuWkWjBvV47qpBXBzbIdBliRyzI14IbWadzOxzM1tpZqlm9idvvLWZfWJma71/W/nNucfM0s1stZld4Dc+xMySvXXTzWvlaGZhZva+N77UzKKr4bWKVJmMPQf4zatLeXh+Gqd0bcPHt56uMJCQV5m/jCkCpjjnegPDgZvNrA9wN/CZcy4G+My7j7duPNAXGAm8aGal19q9BEwCYrzbSG98IrDHOdcdeBqYWgWvTaTKOeeI/2ELI5/5mqSMvfzlsv68/tuhtGuhzyyQ0HfEQ0bOuW3ANm8518xWAh2B0cCZ3mZvAl8Ad3nj7znn8oENZpYODDOzjUAL59wSADN7C7gUWOjNech7rFnA82Zmzjl33K9QpIpk5R7intnJfLYqi5O7tObJsbF0aq021VJ7HNU5BO9QziBgKdDeCwucc9vMrPSTvzsC3/tNy/DGCr3lsuOlc7Z4j1VkZjlAG2BXmeefhG8Pg86dOx9N6SLH5aOkTP48N4WDBcXc/+s+XD8iWm2qpdapdCCYWTNgNnCrc25fBZ/kVN4KV8F4RXN+PuDcDGAGQFxcnPYepNrtPVDA/fNSmZ+YSWxUONPGDaR7O7WpltqpUoFgZg3whcE/nXNzvOEdZhbp7R1EAlneeAbQyW96FJDpjUeVM+4/J8PM6gPhQPYxvB6RKvP5qizump1E9v4CppzXg5vO7KY21VKrVeYqIwNeA1Y6557yW/UhcJ23fB0wz298vHflUBd8J4+XeYeXcs1suPeYE8rMKX2sK4DFOn8ggZKXX8Tds5O4/o0faNWkIfNu+RX/c06MwkBqvcrsIfwKuBZINrMEb+xe4C9AvJlNBDYDYwGcc6lmFg+k4btC6WbnXLE37ybgDaAxvpPJC73x14C3vRPQ2fiuUhKpcUvW7eaOWYlk7j3IjWd047bzYgirr4Z0UjdYqL4Rj4uLc8uXLw90GVJLHCos5q+LVvP6txuIbtOEaeNiGXJS60CXJVLlzGyFcy6uvHX6S2Wp8xK37GVyfALrdu5nwikncfeFvWjSUD8aUvfou17qrIKiEp5bvJYXv1hHu+ZhvDPxZE6NUZtqqbsUCFInrdq+jynxiaRm7uPywVE8cHEfwhurTbXUbQoEqVOKSxwzvlrP05+soUXj+sy4dgjn9z0x0GWJBAUFgtQZG3ftZ8rMRFZs2sOF/U7k0Uv70aZZWKDLEgkaCgSp9UpKHO8s3cQTC1bRoJ7xzJUDGT2wAxX8tb1InaRAkFotc+9B7pyVxDfpuzi9RwR/vXwAJ4arM6lIeRQIUis555jz41Yemp9KcYnjsTH9uHpYZ+0ViFRAgSC1zs7cfO77IJmP03YwLNrXprpzG7WpFjkSBYLUKguTt3Hf3BTy8ou4b1RvfndqF+qpTbVIpSgQpFbIOVDIgx+mMDchk/4dw3lqXCwx7ZsHuiyRkKJAkJD35Zqd3Dkrkd15Bdx6bgw3n9WdBupMKnLUFAgSsvbnF/HYgpW8u3QzMe2a8eqEofSPCg90WSIhS4EgIWnZhmxun5nIlj0HmHR6Vyaf14NGDdSmWuR4KBAkpBwqLGbax6t59ZsNdGrVhPcnncKwLmpTLVIVFAgSMpIy9jI5PpH0rDx+c3Jn7h3Vm6Zh+hYWqSr6aZKgV1hcwvOL03n+83QimoXx5u+GcUaPiECXJVLrKBAkqK3dkcvk+ESSt+YwZlBHHrq4L+FN1KZapDooECQoFZc4XvtmPU9+vIZmYfV5+ZrBjOwXGeiyRGo1BYIEnc27D3D7zESWbczmvD7teXxMfyKaq021SHVTIEjQcM7x7rLNPPbvldQ7wZg2NpbLBndUQzqRGqJAkKCwPecQd85O4qs1Ozktpi1TLx9Ah5aNA12WSJ2iQJCAcs4xN2ErD85LpbDY8b+j+3LN8JO0VyASAAoECZjdefnc90EKi1K3M+SkVkwbG0t026aBLkukzlIgSEB8nLqdez9IZt/BIu6+sBd/OK2r2lSLBJgCQWpUzsFCHp6fypwft9InsgXv/D6WXie2CHRZIoICQWrQN2t3ccesRLJy8/nj2d255ewYGtZXm2qRYHHEn0Yze93MsswsxW/sITPbamYJ3m2U37p7zCzdzFab2QV+40PMLNlbN928s4ZmFmZm73vjS80suopfowTYgYIi7p+bwjWvLaVJw3rMuWkEk8/vqTAQCTKV+Yl8AxhZzvjTzrmB3m0BgJn1AcYDfb05L5pZaU/il4BJQIx3K33MicAe51x34Glg6jG+FglCKzZlM+rZr3ln6SYmntqFf//xNGI7tQx0WSJSjiMGgnPuKyC7ko83GnjPOZfvnNsApAPDzCwSaOGcW+Kcc8BbwKV+c970lmcB55iuOQx5+UXFPLFwJWNfXkJRieNffxjO/b/uo88sEAlix3MO4RYzmwAsB6Y45/YAHYHv/bbJ8MYKveWy43j/bgFwzhWZWQ7QBthV9gnNbBK+vQw6d+58HKVLdUrZmsOU+ERW78jlqmGduO+iPjRTm2qRoHesB3FfAroBA4FtwDRvvLx39q6C8Yrm/HLQuRnOuTjnXFxEhNofB5ui4hKe+2wtl77wLXsOFPCP3w7licsGKAxEQsQx/aQ653aULpvZ34GPvLsZQCe/TaOATG88qpxx/zkZZlYfCKfyh6gkSKRn5TElPoHEjBwuie3AI6P70rJJw0CXJSJH4Zj2ELxzAqXGAKVXIH0IjPeuHOqC7+TxMufcNiDXzIZ75wcmAPP85lznLV8BLPbOM0gIKClxvPr1ei6a/jWbsw/wwtWDmX7VIIWBSAg64h6Cmf0LOBNoa2YZwIPAmWY2EN+hnY3ADQDOuVQziwfSgCLgZudcsfdQN+G7YqkxsNC7AbwGvG1m6fj2DMZXweuSGrAl29emeumGbM7t3Y7HL+tPu+aNAl2WiBwjC9U343FxcW758uWBLqNOcs7x3g9bePSjNMyMBy7uw9ghUWpIJxICzGyFcy6uvHU62ydHZce+Q9w9O4nPV+/klK5t+NvYAUS1ahLoskSkCigQpNI+TMzk/rkp5BcV89DFfZhwSjQnqCGdSK2hQJAjyt5fwP3zUvh30jYGdW7JtLGxdI1oFuiyRKSKKRCkQp+m7eDuOcnkHCzgjgt6csPpXalfTz2IRGojBYKUK/dQIY/MT2Pmigx6ndictycOo3ek2lSL1GYKBPmF79J3ccesJLblHOTms7rxp3N6qDOpSB2gQJCfHCwoZuqiVbzx3Ua6tm3KrJtGMLhzq0CXJSI1RIEgAPy4eQ+3xyeyftd+fjsimrtG9qJxQ3UmFalLFAh1XH5RMdM/W8tLX6wjMrwx7/7+ZEZ0bxvoskQkABQIdVha5j4mxyewansu4+KiuP/XfWjeqEGgyxKRAFEg1EFFxSW88tV6nvl0DeGNG/LqhDjO7dM+0GWJSIApEOqY9TvzmDIzkf9u3stFAyJ5dHQ/WjVVZ1IRUSDUGSUljreWbOQvi1YRVr8e068axCWxHQJdlogEEQVCHZCx5wB3zkriu3W7ObNnBFMvH0D7FmpTLSI/p0CoxZxzzFyRwSPz03DO8ZfL+nPl0E5qUy0i5VIg1FJZuYe4Z3Yyn63K4uQurXlybCydWqtNtYgcngKhFvooKZM/z03hYEEx9/+6D9ePUJtqETkyBUItsvdAAffPS2V+YiaxUeFMGzeQ7u3UplpEKkeBUEt8viqLu2Ynkb2/gCnn9eCmM7upTbWIHBUFQojLyy/isX+n8a9lW+jZvjmv/3Yo/TqGB7osEQlBCoQQ9v363dw+M5HMvQe58Yxu3HZeDGH11ZBORI6NAiEEHSos5q+LVvP6txuIbtOEmTeewpCTWge6LBEJcQqEEJO4ZS+T4xNYt3M/E045ibsv7EWThvpvFJHjp98kIaKgqITnFq/lxS/W0a55GO9MPJlTY9SmWkSqjgIhBKzensvk+ARSM/dx+eAoHri4D+GN1aZaRKqWAiGIFZc4/v71ep76eA0tGtfnlWuHcEHfEwNdlojUUke8UN3MXjezLDNL8RtrbWafmNla799WfuvuMbN0M1ttZhf4jQ8xs2Rv3XTzGuqYWZiZve+NLzWz6Cp+jSFp4679jHtlCX9ZuIqze7XjP7eerjAQkWpVmb9cegMYWWbsbuAz51wM8Jl3HzPrA4wH+npzXjSz0usgXwImATHerfQxJwJ7nHPdgaeBqcf6YmqD0jbVFz77NWt35PLMlQN56ZrBtGkWFujSRKSWO2IgOOe+ArLLDI8G3vSW3wQu9Rt/zzmX75zbAKQDw8wsEmjhnFvinHPAW2XmlD7WLOAcq6PtODP3HmTC68t4YF4qQ7u05uPbzuDSQR3VnVREasSxnkNo75zbBuCc22Zm7bzxjsD3fttleGOF3nLZ8dI5W7zHKjKzHKANsKvsk5rZJHx7GXTu3PkYSw8+zjnm/LiVh+anUlzieGxMP64e1llBICI1qqpPKpf3G8xVMF7RnF8OOjcDmAEQFxdX7jahZldePvfOSebjtB0MjW7Fk2NjOalN00CXJSJ10LEGwg4zi/T2DiKBLG88A+jkt10UkOmNR5Uz7j8nw8zqA+H88hBVrbQoZRv3fpBCXn4R943qze9O7UI9takWkQA51naYHwLXecvXAfP8xsd7Vw51wXfyeJl3eCnXzIZ75wcmlJlT+lhXAIu98wy1Vs6BQm57P4Eb3/mRDi0b8dH/nMofTu+qMBCRgDriHoKZ/Qs4E2hrZhnAg8BfgHgzmwhsBsYCOOdSzSweSAOKgJudc8XeQ92E74qlxsBC7wbwGvC2maXj2zMYXyWvLEh9uWYnd85KZHdeAbeeG8PNZ3WngdpUi0gQsFB9Mx4XF+eWL18e6DIqbX9+EY8tWMm7SzcT064ZT40bSP8otakWkZplZiucc3HlrdNfKteAZRuyuX1mIlv2HGDS6V2ZfF4PGjVQm2oRCS4KhGp0qLCYpz5Zw9+/Xk+nVk14f9IpDOuiNtUiEpwUCNUkOSOHyfEJrM3K4zcnd+beUb1pGqYvt4gEL/2GqmKFxSW88Hk6zy9Op02zhrz5u2Gc0SMi0GWJiByRAqEKrd2Ry+T4RJK35jBmUEceurgv4U3UplpEQoMC4TjlHCxk/c48vk3fxfTF6TQLq8/L1wxmZL/IQJcmInJUFAiVtDsvnzU78liblcvq7bmszcpj/c797MrL/2mb8/q05/Ex/Ylors6kIhJ6FAjlKCwuISkjhyXrdrF0QzYrt+1jV17BT+vDGzege7tmnN0rgq4RzegW0Yzu7ZoR3aaJGtKJSMhSIOD7ZLK0zH0sWb+L79bt5ocN2ewv8P2Bda8Tm3N2r3b0aN/8p1v7FmH6xS8itU6dDATnHGt25PHdul0sWbeb79fvZt+hIgC6RTTlssFRnNKtDcO7tqF104YBrlZEpGbUuUB4b9lmnvx49U+HgDq1bsyF/SIZ0d0XAO1bNApwhSIigVHnAqF9eCNOi4nglG5tOKVrGzq1bhLokkREgkKdC4SzerbjrJ7tjryhiEgdo77LIiICKBBERMSjQBAREUCBICIiHgWCiIgACgQREfEoEEREBFAgiIiIx5xzga7hmJjZTmBTDT1dW2BXDT1XVVLdNScUawbVXZOCpeaTnHPlfoxjyAZCTTKz5c65uEDXcbRUd80JxZpBddekUKhZh4xERARQIIiIiEeBUDkzAl3AMVLdNScUawbVXZOCvmadQxAREUB7CCIi4lEgiIgIUAsDwcxGmtlqM0s3s7v9xt83swTvttHMEg4zf6yZpZpZiZnF+Y2fZ2YrzCzZ+/fsw8z/p/f8KWb2upk18MZ/Y2ZJ3u07M4utgZqH+c1PNLMxh5nfxcyWmtla7zkbeuNmZtO9upLMbHANfa2Pt+7Dfq2DuW6/9UPNrNjMrgiFms3sTG9+qpl9WWZeUNZtZuFmNt+bm2pm1wdZ3bd4z+3MrK3fuFkFP5PHzTlXa25APWAd0BVoCCQCfcrZbhrwwGEeozfQE/gCiPMbHwR08Jb7AVsPM38UYN7tX8BN3vgIoJW3fCGwtAZqbgLU95YjgazS+2XmxwPjveWX/WoeBSz0Xsvw0ppDoO5yv9bBXrdffYuBBcAVwV4z0BJIAzp799uFwtcauBeY6i1HANlAwyCqexAQDWwE2pb5/VLuz2RV3GrbHsIwIN05t945VwC8B4z238DMDBiH75f1LzjnVjrnVpcz/l/nXKZ3NxVoZGZh5Wy3wHmAZUCUN/6dc26Pt9n3pePVXPMB51yRd7cR8IsrCLzHPhuY5Q29CVzqLY8G3vJezvdASzOLDPa6K/haB3Xdnv8BZuP7RREKNV8NzHHObfYeL1TqdkBzb5tm+AKh9PECWre33X+dcxvLWVXRz+Rxq22B0BHY4nc/wxvzdxqwwzm39jie53Lgv865/MNtYL5DRdcCi8pZPRFfykM112xmJ5tZKpAM3Oj3zViqDbDXb9z/+SuqLZjr9uf/tQ7qus2sIzAG3ztZf0FbM9ADaGVmX5jvUOqEEKn7eXzv4jO9+X9yzpUESd0VqUxtx6y2BYKVM1Y2ga/iMKleqScw6wtMBW44wqYvAl85574uM/8sfL+k7iodKmduldXsnFvqnOsLDAXuMbNGZTap6PmPdV2pQNXt2+CXX+tKzSNwdT8D3OWcKz6KOaUCVXN9YAhwEXABcL+Z9QiBui8AEoAOwEDgeTNrUYl5paqz7opUprZjVtsCIQPo5Hc/Ct87AADMrD5wGfC+39g/vBM8C4704GYWBXwATHDOratguwfxHZecXGZ8APAqMNo5t7smai7lnFsJ7Md3/sPfLny7nfXLef6Kagvmug/3tQ72uuOA98xsI3AF8KKZXRrkNWcAi5xz+51zu4CvgFi/dcFa9/X4DnU551w6sAHoFSR1V6TC2o5beScWQvWG793KeqAL/3cyqK/f+pHAl5V8rC/4+cmglt7jXX6Eeb8HvgMalxnvDKQDI2qw5i783wmsk7xvnLblzJvJz0+8/T9v+SJ+fgJrWYjUXe7XOtjrLrPNG/zfSeWgrRnfYZfPvBqbAClAvxCo+yXgIW+5PbC1dH4w1O23/UZ+flL5sD+TVXGrsgcKlhu+s/Br8F0lcF+ZdW/gO2ZX0fwx+FI4H9gB/Mcb/zO+NE/wu7UrZ36R99yl2zzgjb8K7PEbX14DNV+L7wR4AvAjcOlh5nfFdwI83fsBCvPGDXjBqyvZ/xs7yOs+7Nc6mOsup44rQqFm4A58VxqlALeGwtca36Gij/F9X6cA1wRZ3X/05hfhC41XK/Mzebw3ta4QERGg9p1DEBGRY6RAEBERQIEgIiIeBYKIiAAKBBER8SgQREQEUCCIiIjn/wNFgRHbQCKl/gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to run this section\n",
    "run_timing_check = False\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "old_entries = {\n",
    "    \"Jul 12 2021 7:02PM\" : 3696,\n",
    "    \"Jul 12 2021 7:43PM\" : 4843,\n",
    "    \"Jul 12 2021 8:27PM\" : 16821,\n",
    "    \"Jul 12 2021 8:51PM\" : 21158,\n",
    "    \"Jul 12 2021 9:28PM\" : 25486\n",
    "}\n",
    "\n",
    "entries = {\n",
    "    \"Jul 12 2021 9:29PM\" : 'dasgg_TRINITY_GG_110_c20_g1_i3.mrna1',\n",
    "    \"Jul 12 2021 9:59PM\" : 's13_TRINITY_DN50407_c9_g1_i1.mrna1',\n",
    "    \"Jul 12 2021 10:08PM\" : 's13_TRINITY_DN5591_c0_g1_i1.mrna1',\n",
    "    \"Jul 12 2021 10:13PM\" : 'kao_mikado.phaw_50.000081aG112.1.mrna1',\n",
    "    \"Jul 12 2021 10:25PM\" : 'st2m_STRG.116.3',\n",
    "    \"Jul 13 2021 10:48AM\" : 'das_TRINITY_DN162887_c3_g2_i8.mrna1'\n",
    "}\n",
    "\n",
    "total_genes = 3959963\n",
    "\n",
    "if run_timing_check == True:\n",
    "    \n",
    "    times = []\n",
    "    numbers = []\n",
    "    \n",
    "    for key in entries:\n",
    "        times = times + [datetime.datetime.strptime(key, '%b %d %Y %I:%M%p')]\n",
    "        transcript_string = entries[key]\n",
    "        \n",
    "        old_stdout = sys.stdout\n",
    "        sys.stdout = mystdout = StringIO()\n",
    "\n",
    "        !awk '/'{transcript_string}'/{{ print NR; exit }}' mikado_prepared.fasta\n",
    "\n",
    "        sys.stdout = old_stdout\n",
    "        linenum = int(mystdout.getvalue().split(\"\\r\")[0])\n",
    "        \n",
    "        old_stdout = sys.stdout\n",
    "        sys.stdout = mystdout = StringIO()\n",
    "        \n",
    "        !head -n {linenum} mikado_prepared.fasta | grep -c '>'\n",
    "        \n",
    "        sys.stdout = old_stdout\n",
    "        returned = int(mystdout.getvalue().split(\"\\r\")[0])\n",
    "        \n",
    "        numbers = numbers + [returned]\n",
    "    \n",
    "    #calculates total seconds from last to first entry\n",
    "    time_diff = (times[-1] - times[0]).total_seconds()\n",
    "    \n",
    "    #calculates total number of genes output from last to first entry\n",
    "    number_diff = (numbers[-1] - numbers[0])\n",
    "    \n",
    "    #calculates genes per second\n",
    "    rate = number_diff / time_diff\n",
    "    display('approximately ' + str(rate) + ' genes per second')\n",
    "    \n",
    "    remain_genes = total_genes - numbers[-1]\n",
    "    \n",
    "    display('approximately ' + str(remain_genes) + ' genes remain')\n",
    "    \n",
    "    remain_time = remain_genes / rate\n",
    "    finish_time = times[-1] + datetime.timedelta(seconds=remain_time)\n",
    "    display(remain_time / 60 / 60 / 24)\n",
    "    display('expected finish at ' + str(finish_time))\n",
    "    plt.plot(times, numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to run this section\n",
    "run_trinotate_prepare = False\n",
    "run_transdecoder_mikado = True\n",
    "run_trinotate_blast = False\n",
    "run_trinotate_hmmscan = False\n",
    "run_trinotate_signalp = False\n",
    "run_trinotate_tmhmm = False\n",
    "run_RNAMMER = False\n",
    "\n",
    "mikado_loci_fasta = 'DAS.mikado.loci.fasta'\n",
    "transdecoder_name = mikado_loci_fasta + \".transdecoder.pep\"\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "if run_trinotate_prepare == True:\n",
    "    #before running this, make sure you have loaded the modules using:\n",
    "    #module load trinotate perl-dbi\n",
    "    !/clusterfs/vector/home/groups/software/sl-7.x86_64/modules/trinotate/3.1.1/admin/Build_Trinotate_Boilerplate_SQLite_db.pl Trinotate\n",
    "    !makeblastdb -in uniprot_sprot.pep -dbtype prot\n",
    "    !gunzip Pfam-A.hmm.gz\n",
    "    !hmmpress Pfam-A.hmm\n",
    "\n",
    "if run_transdecoder_mikado == True:\n",
    "    #make sure you have loaded transdecoder\n",
    "    #module load transdecoder\n",
    "    !TransDecoder.LongOrfs -t $mikado_loci_fasta && TransDecoder.Predict -t $mikado_loci_fasta\n",
    "    \n",
    "if run_trinotate_blast == True:\n",
    "    !blastx -query $mikado_loci_fasta -db uniprot_sprot.pep -num_threads 12 -max_target_seqs 1 -outfmt 6 -evalue 1e-3 > blastx.outfmt6\n",
    "    !blastp -query $transdecoder_name -db uniprot_sprot.pep -num_threads 12 -max_target_seqs 1 -outfmt 6 -evalue 1e-3 > blastp.outfmt6\n",
    "\n",
    "if run_trinotate_hmmscan == True:\n",
    "    !hmmscan --cpu 12 --domtblout TrinotatePFAM.out Pfam-A.hmm $transdecoder_name > pfam.log\n",
    "\n",
    "if run_trinotate_signalp == True:\n",
    "    !signalp -f short -n signalp.out $transdecoder_name\n",
    "\n",
    "if run_trinotate_tmhmm == True:\n",
    "    !tmhmm --short < transdecoder.pep > tmhmm.out\n",
    "\n",
    "if run_RNAMMER == True:\n",
    "    !/clusterfs/vector/home/groups/software/sl-7.x86_64/modules/trinotate/3.1.1/util/rnammer_support/RnammerTranscriptome.pl --transcriptome $mikado_loci_fasta --path_to_rnammer /usr/bin/software/rnammer_v1.2/rnammer\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
