{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "125382d5-8335-4847-8674-7ef5994f6c4d",
   "metadata": {},
   "source": [
    "# Setup\n",
    "0. added the following lines to`data.config.user` and moved all the files to the right place\n",
    "\n",
    "```\n",
    "[ph5]\n",
    "genome: ph5/phaw_5.0.fa\n",
    "chromosome_sizes: ph5/phaw_5.0.chrom.sizes\n",
    "gene_regions: ph5/mikado.loci.sorted.transcripts.bed\n",
    "annotation: ph5/mikado.loci.sorted.gtf\n",
    "gene_alias: ph5/DAS.mikado.loci.fasta.gene_trans_map\n",
    "```\n",
    "\n",
    "1. downloaded all non-redundant CORE JASPAR sequences as .jaspar files from <http://jaspar.genereg.net/downloads/>\n",
    "2. Used the R script `prepare_for_rgt.Rmd` to convert all JASPAR sequences to the format RGT likes\n",
    "3. `prepare_for_rgt.Rmd` also generates a nice .mtf file for us to use\n",
    "4. upload to correct directory in `~/rgtdata/motifs`\n",
    "5. move the .mtf file to the right place\n",
    "\n",
    "```\n",
    "mv jaspar_core_nr.mtf ../\n",
    "```\n",
    "\n",
    "6. go back to `~/rgtdata` and run this:\n",
    "\n",
    "```\n",
    "python setupLogoData.py --all\n",
    "```\n",
    "\n",
    "7. modify `data.config.user` to add a line specifying `jaspar_core_nr` as default database\n",
    "```\n",
    "repositories: jaspar_core_nr\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ecc124-bad3-4a89-a950-c6a417a11451",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to make a sample file\n",
    "run_hintatac_footprint = False\n",
    "run_hintatac_tracks_rerun = False\n",
    "run_hintatac_wigtobigwig = False\n",
    "run_hintatac_motifmatch = False\n",
    "run_hintatac_motifmatch_random = False\n",
    "run_hintatac_motifmatch_manualrandom = False\n",
    "run_hintatac_enrichment = False\n",
    "run_hintatac_motifmatch_manualrandom_smaller = False\n",
    "run_hintatac_enrichment_smaller = False\n",
    "run_hintatac_enrichment_smaller_all = False\n",
    "\n",
    "stage_array = ['13', '14', '15', '17', '18', '19', '19plus', '20', '21', '22', '23', '24', '25', '26', '27']\n",
    "bam_pattern = '*AB_bothruns_q10.Genrich_resorted.bam'\n",
    "narrowPeak_pattern = '*A_bothruns_q10.Genrich_sorted.ATAC.q005.narrowPeak'\n",
    "allpeaks_bed = 'OmniATAC_bothruns_q005_allpeaks.igv_new.bed'\n",
    "\n",
    "match_specific = 'match_specific'\n",
    "enrichment_specific = 'enrichment_specific'\n",
    "match_all = 'match_all'\n",
    "enrichment_all = 'enrichment_all'\n",
    "manualrandom_specific = 'random_regions_forspecific.bed'\n",
    "manualrandom_all = 'random_regions_forall.bed'\n",
    "\n",
    "enrichment_specific_small = 'enrichment_specific_small'\n",
    "enrichment_all_small = 'enrichment_all_small'\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "if run_hintatac_footprint:\n",
    "    for stage in stage_array:\n",
    "        stage_prefix = 'S' + stage \n",
    "        stage_bam = bam_pattern.replace('*', stage_prefix)\n",
    "        stage_abbrev = stage_prefix + '_hintatac'\n",
    "        stage_narrowPeak = 'Genrich_sorted_bothruns/q005_narrowPeak/' + narrowPeak_pattern.replace('*', stage_prefix)\n",
    "        \n",
    "        !echo '### performing stage specific analysis of '{stage_prefix}'\\n'\n",
    "        stage_specific = stage_abbrev + '_specific'\n",
    "        !rgt-hint footprinting --atac-seq --paired-end --organism=ph5 --output-location=./hintatac --output-prefix={stage_specific} {stage_bam} {stage_narrowPeak}\n",
    "        \n",
    "        stage_specific_bc = stage_specific + '_bc'\n",
    "        !rgt-hint tracks --bc --bigWig --organism=ph5 {stage_bam} {stage_narrowPeak} --output-prefix={stage_specific_bc}\n",
    "        \n",
    "        !echo '### performing allpeaks analysis of '{stage_prefix}'\\n'\n",
    "        stage_all = stage_abbrev + '_all'\n",
    "        !rgt-hint footprinting --atac-seq --paired-end --organism=ph5 --output-location=./hintatac --output-prefix={stage_all} {stage_bam} {allpeaks_bed}\n",
    "        \n",
    "        stage_all_bc = stage_all + '_bc'\n",
    "        !rgt-hint tracks --bc --bigWig --organism=ph5 {stage_bam} {allpeaks_bed} --output-prefix={stage_all_bc}\n",
    "\n",
    "if run_hintatac_tracks_rerun:\n",
    "    #used vim /global/home/users/dasun/.local/lib/python3.7/site-packages/rgt/HINT/signalProcessing.py\n",
    "    #to comment out the for loop beginning with \"for read in self.bam.fetch\"\n",
    "    #as per this GitHub thread: https://github.com/CostaLab/reg-gen/issues/162\n",
    "    #also adjusted so that if p1, p2, p1_w, or p2_w are <= 0, sets area to 0\n",
    "    for stage in stage_array:\n",
    "        stage_prefix = 'S' + stage \n",
    "        stage_bam = bam_pattern.replace('*', stage_prefix)\n",
    "        stage_abbrev = stage_prefix + '_hintatac'\n",
    "        stage_narrowPeak = 'Genrich_sorted_bothruns/q005_narrowPeak/' + narrowPeak_pattern.replace('*', stage_prefix)\n",
    "        \n",
    "        !echo '### performing stage specific analysis of '{stage_prefix}'\\n'\n",
    "        stage_specific = stage_abbrev + '_specific'\n",
    "        stage_specific_bc = stage_specific + '_bc'\n",
    "        !rgt-hint tracks --bc --organism=ph5 {stage_bam} {stage_narrowPeak} --output-prefix={stage_specific_bc}\n",
    "        \n",
    "        !echo '### performing allpeaks analysis of '{stage_prefix}'\\n'\n",
    "        stage_all = stage_abbrev + '_all'\n",
    "        stage_all_bc = stage_all + '_bc'\n",
    "        !rgt-hint tracks --bc --organism=ph5 {stage_bam} {allpeaks_bed} --output-prefix={stage_all_bc}\n",
    "\n",
    "if run_hintatac_wigtobigwig:\n",
    "    #had to run this separately because the current install via HINT-ATAC for wigToBigWig is somehow broken\n",
    "    #downloaded wigToBigWig from UCSC using wget http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/wigToBigWig\n",
    "    #chmod +x wigToBigWig\n",
    "    for stage in stage_array:\n",
    "        all_name = 'S' + stage + '_hintatac_all_bc.wig'\n",
    "        all_bw_name = all_name.replace('.wig', '.bw')\n",
    "        !/global/scratch/dasun/wigToBigWig {all_name} phaw_5.0.chrom.sizes {all_bw_name}\n",
    "        \n",
    "        specific_name = 'S' + stage + '_hintatac_specific_bc.wig'\n",
    "        specific_bw_name = specific_name.replace('.wig', '.bw')\n",
    "        !/global/scratch/dasun/wigToBigWig {specific_name} phaw_5.0.chrom.sizes {specific_bw_name}\n",
    "        \n",
    "if run_hintatac_motifmatch:\n",
    "    specific_bed_array = ['./hintatac/S' + stage + '_hintatac_specific.bed' for stage in stage_array]\n",
    "    specific_bed_string = ' '.join(specific_bed_array)\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --input-files {specific_bed_string}\n",
    "    \n",
    "    all_bed_array = ['./hintatac/S' + stage + '_hintatac_all.bed' for stage in stage_array]\n",
    "    all_bed_string = ' '.join(all_bed_array)\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --input-files {all_bed_string}\n",
    "\n",
    "if run_hintatac_motifmatch_random:\n",
    "    specific_bed_array = ['./hintatac/S' + stage + '_hintatac_specific.bed' for stage in stage_array]\n",
    "    specific_bed_string = ' '.join(specific_bed_array)\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --rand-proportion 5 --output-location {match_specific} --input-files {specific_bed_string} \n",
    "    \n",
    "    all_bed_array = ['./hintatac/S' + stage + '_hintatac_all.bed' for stage in stage_array]\n",
    "    all_bed_string = ' '.join(all_bed_array)\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --rand-proportion 5 --output-location {match_all} --input-files {all_bed_string} \n",
    "\n",
    "if run_hintatac_motifmatch_manualrandom:\n",
    "    !echo '### running motif matching on random peaks from '{manualrandom_specific}'\\n'\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --output-location {match_specific} --input-files {manualrandom_specific}\n",
    "    !echo '### running motif matching on random peaks from '{manualrandom_all}'\\n'\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --output-location {match_all} --input-files {manualrandom_all}\n",
    "    \n",
    "if run_hintatac_enrichment:\n",
    "    specific_bed_array = ['./hintatac/S' + stage + '_hintatac_specific.bed' for stage in stage_array]\n",
    "    specific_bed_string = ' '.join(specific_bed_array)\n",
    "    !rgt-motifanalysis enrichment --organism=ph5 --matching-location {match_specific} --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --output-location {enrichment_specific} random_regions_forspecific.bed {specific_bed_string}\n",
    "    \n",
    "    all_bed_array = ['./hintatac/S' + stage + '_hintatac_all.bed' for stage in stage_array]\n",
    "    all_bed_string = ' '.join(all_bed_array)\n",
    "    !rgt-motifanalysis enrichment --organism=ph5 --matching-location {match_all} --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --output-location {enrichment_all} random_regions_forspecific.bed {all_bed_string}\n",
    "    \n",
    "if run_hintatac_motifmatch_manualrandom_smaller:\n",
    "    !echo '### running motif matching on random peaks from random_regions_smaller.bed\\n'\n",
    "    !rgt-motifanalysis matching --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --organism=ph5 --output-location {match_specific} --input-files random_regions_smaller.bed\n",
    "    !cp match_specific/random_regions_smaller_mbps.bed match_all/random_regions_smaller_mbps.bed\n",
    "    \n",
    "if run_hintatac_enrichment_smaller:\n",
    "    specific_bed_array = ['./hintatac/S' + stage + '_hintatac_specific.bed' for stage in stage_array]\n",
    "    specific_bed_string = ' '.join(specific_bed_array)\n",
    "    !rgt-motifanalysis enrichment --organism=ph5 --matching-location {match_specific} --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --output-location {enrichment_specific_small} random_regions_smaller.bed {specific_bed_string}\n",
    "\n",
    "if run_hintatac_enrichment_smaller_all:\n",
    "    all_bed_array = ['./hintatac/S' + stage + '_hintatac_all.bed' for stage in stage_array]\n",
    "    all_bed_string = ' '.join(all_bed_array)\n",
    "    !rgt-motifanalysis enrichment --organism=ph5 --matching-location {match_all} --motif-dbs ~/rgtdata/motifs/jaspar_core_nr --output-location {enrichment_all_small} random_regions_smaller.bed {all_bed_string}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f91e89-7797-42f4-8fd3-a9c1bcebee53",
   "metadata": {},
   "outputs": [],
   "source": [
    "#############################################################\n",
    "### Fill in this information before running this section ####\n",
    "\n",
    "#Decide if you want to make a sample file\n",
    "run_hintatac_split_footprints = False\n",
    "run_hintatac_split_mbps = False\n",
    "\n",
    "#Set number of acores\n",
    "acores = 9\n",
    "\n",
    "hintatac_dir = 'hintatac/'\n",
    "hintatac_split_dir = 'hintatac_9acoresplit/'\n",
    "hintatac_footprint_pattern = '_hintatac_all.bed'\n",
    "\n",
    "match_dir = 'match_all/'\n",
    "match_split_dir = 'match_all_9acoresplit/'\n",
    "match_pattern = '_hintatac_all_mpbs.bed'\n",
    "\n",
    "stage_array = ['13', '14', '15', '17', '18', '19', '19plus', '20', '21', '22', '23', '24', '25', '26', '27']\n",
    "Sstage_array = ['S' + i for i in stage_array]\n",
    "\n",
    "Mfuzz_acorebed_pattern = 'Mfuzz_' + str(acores) + 'acores_acore*_*_peaks.bed'\n",
    "\n",
    "#############################################################\n",
    "#############################################################\n",
    "\n",
    "if run_hintatac_split_footprints:\n",
    "    #Generate string of all hintatac footprint bed file names\n",
    "    hintatac_footprint_list = [hintatac_dir + stage + hintatac_footprint_pattern for stage in Sstage_array]\n",
    "    hintatac_footprint_string = ' '.join(hintatac_footprint_list)\n",
    "    \n",
    "    !printf \"Collecting overlaps for all footprint files listed: \"{hintatac_footprint_string}\"\\n\"\n",
    "    \n",
    "    #Iterate through number of acores\n",
    "    for acore_num in np.arange(1, acores + 1):\n",
    "        \n",
    "        #Generate names for files\n",
    "        Mfuzz_acorebed = Mfuzz_acorebed_pattern.replace('*_*', str(acore_num))\n",
    "        splitbed_filename = hintatac_split_dir + 'Sall_hintatac_all_acore' + str(acore_num) + '.bed'\n",
    "        \n",
    "        #Bedtools intersect each individual acore bed file vs. all footprint addresses\n",
    "        !bedtools intersect -a {Mfuzz_acorebed} -b {hintatac_footprint_string} -sorted -wb > {splitbed_filename}\n",
    "        \n",
    "        #Use sort to remove non-unique coordinates, output to a new file\n",
    "        splitbed_uniquename = splitbed_filename.replace('.bed', '.unique.bed')\n",
    "        !printf \"removing lines from \"{splitbed_filename}\" with non-unique coordinates and output to \"{splitbed_uniquename}\n",
    "        !sort -k1,1 -k2,2n -k3,3n -u {splitbed_filename} > {splitbed_uniquename}\n",
    "\n",
    "if run_hintatac_split_mbps:\n",
    "    #Generate string of all hintatac footprint bed file names\n",
    "    match_list = [match_dir + stage + match_pattern for stage in Sstage_array]\n",
    "    match_string = ' '.join(match_list)\n",
    "    \n",
    "    !printf \"Collecting overlaps for all mbps files listed: \"{match_string}\"\\n\"\n",
    "    \n",
    "    #Iterate through number of acores\n",
    "    for acore_num in np.arange(1, acores + 1):\n",
    "        \n",
    "        #Generate names for files\n",
    "        Mfuzz_acorebed = Mfuzz_acorebed_pattern.replace('*_*', str(acore_num))\n",
    "        splitbed_filename = match_split_dir + 'Sall_hintatac_all_acore' + str(acore_num) + '_mbps.bed'\n",
    "        \n",
    "        #Bedtools intersect each individual acore bed file vs. all footprint addresses\n",
    "        !bedtools intersect -a {Mfuzz_acorebed} -b {match_string} -sorted -wb > {splitbed_filename}\n",
    "        \n",
    "        #Use sort to remove non-unique coordinates, output to a new file\n",
    "        #Also keep any matched lines\n",
    "        splitbed_uniquename = splitbed_filename.replace('.bed', '.unique.bed')\n",
    "        !printf \"removing lines from \"{splitbed_filename}\" with non-unique entries and output to \"{splitbed_uniquename}\n",
    "        !sort -k1,1 -k2,2n -k3,3n, -k4,4n, -k5,5n -u {splitbed_filename} > {splitbed_uniquename}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
