# GEO Files
This file describes all of the files found in the Gene Expression Omnibus (GEO) submissions associated with the Sun et al. 2022 manuscript.


## directory `ImpulseDE2/`
This directory contains ImpulseDE2 and DESeq2 differential expression analysis files.

| filename | description |
|-------------:|:---------------|
| `OmniATAC-Seq_bothruns_ImpulseDE2_ImpulseParams_thpf.csv` | Impulse Parameters extracted from ImpulseDE2 analysis of all developmental stages. _Used in ParhyaleATACExplorer scripts._ |
| `OmniATAC-Seq_bothruns_ImpulseDE2_ImpulseParams_thpf_noS21S22.csv` | Impulse Parameters extracted from ImpulseDE2 analysis of all developmental stages except for stages S21 and S22, used for peak stability analyses. _Used in ParhyaleATACExplorer scripts._ |
| `OmniATAC-Seq_bothruns_ImpulseDE2_modelfits_thpf.csv` | IDE2 model fits extracted from ImpulseDE2 analysis of all developmental stages. |
| `OmniATAC-Seq_bothruns_ImpulseDE2_modelfits_thpf_noS21S22.csv` | IDE2 model fits extracted from ImpulseDE2 analysis of all developmental stages except for stages S21 and S22. |
| `OmniATAC-Seq_bothruns_ImpulseDE2_sigstable_max.tsv` | Max accessibility timing for IDE2 model fits that were considered statistically significantly changing by IDE2. |
| `OmniATAC-Seq_bothruns_ImpulseDE2_sigstable_stable_thpf.tsv` | IDE2 model fits for significantly changing peaks that were also stable to loss of S21/S22 stages. |
| `OmniATAC-Seq_bothruns_ImpulseDE2_thpf.tsv` | Differential accessibilty measurements for ImpulseDE2 for all ata. Includes significance metrics. |
| `OmniATAC-Seq_bothruns_ImpulseDE2_thpf_noS21S22.tsv` | Differential accessibilty measurements for ImpulseDE2 for all data except S21/S22 stages. Includes significance metrics. |
| `OmniATAC_ImpulseDE2_sample_matrix_time_is_hpf.tsv` | ImpulseDE2 sample file matrix for IDE2 analysis. |
| `OmniATAC_bothruns_multicov_forDESeq.tsv` | Multicov accessibility values for merged peaks across all libraries. _Used in ParhyaleATACExplorer scripts._ |
| `OmniATAC_bothruns_multicov_mfuzz_mikado.tsv` | Multicov accessibility values for use in Mfuzz analyses. |
| `OmniATAC_bothruns_multicov_mfuzz_mikado_standardised.tsv` | Multicov accessibility values, row-standardized by Mfuzz. _Used in ParhyaleATACExplorer scripts._ |
| `OmniATAC_bothruns_multicov_unformat_mikado.tsv` | Multicov accessibility values, unformatted. _Used in ParhyhaleATACExplorer scripts._|
| `S19_vs_S13_OmniATACSeq_results.csv` | DESeq2 differential accessibility fold change and padj values for concordance/discordance analyses. _Used in ParhyhaleATACExplorer scripts._|
| `S21_vs_S19_OmniATACSeq_results.csv` | DESeq2 differential accessibility fold change and padj values for concordance/discordance analyses. _Used in ParhyhaleATACExplorer scripts._|
| `S21_vs_S23_OmniATACSeq_results.csv` | DESeq2 differential accessibility fold change and padj values for concordance/discordance analyses. _Used in ParhyhaleATACExplorer scripts._|




## directory `Mfuzz_mikado/`
This directory contains information about the clustering analysis from Mfuzz.

___NOTE:___ Each "acore" assigned by Mfuzz is NOT identical to each Cluster referred to in the manuscript. The following table explains the relationship between each "acore" in these files and the final Clusters in the paper:

| acore number | Cluster number |
|-------------:|:---------------|
| acore6 | Cluster 1 |
| acore9 | Cluster 2 |
| acore5 | Cluster 3 |
| acore8 | Cluster 4 |
| acore3 | Cluster 5 |
| acore7 | Cluster 6 |
| acore2 | Cluster 7 |
| acore1 | Cluster 8 |
| acore4 | Cluster 9 |

#### `acore#_peaks.bed` files
Peaks for each acore.

#### `Mfuzz_a9_m1.1_clusters.tsv` file
TSV file containing all peaks, and the membership score for each peak for each acore.

#### `Mfuzz_a9_m1.1_cores.txt` file
Text file of all acores combined, including membership value for each acore.

#### `a9_m1.1_cores_#acore.txt` files
Peak addresses and membership value for each acore.

#### `Mfuzz_a9_m1.1_overlap.txt` file
Acore overlap matrix.




## directory `NucATAC_bedgz/`
This directory contains .bed.gz files for data generated by NucleoATAC. For more details, see the [NucleoATAC documentation](https://nucleoatac.readthedocs.io/en/latest/nucleoatac/).

#### `.nfrpos.bed.gz` files
Nucleosome-free regions for each stage, as called by NucleoATAC. _Use these data for visualizing NFRs._

#### `nucmap_combined.bed.gz` files
Combined nucleosome positions (merging .nucpos. and .occpeaks. files) for each stage, as called by NucleoATAC. _Use these data for visualizing nucleosome positions._

#### `.nucpos.bed.gz` files
More precise nucleosome positions for each stage.

#### `.occpeaks.bed.gz` files
Crude nucleosome occupancy for each stage, as called by NucleoATAC.




## directory `NucATAC_bw/`
This directory contains bigwig files for data generated by NucleoATAC. For more details, see the [NucleoATAC documentation](https://nucleoatac.readthedocs.io/en/latest/nucleoatac/). 

#### `.ins.bw` files
Insertion site pileups in bigwig format for each developmental stage.

#### `.nucleoatac_signal.bw` files
Raw NucleoATAC signal for each developmental stage.

#### `.nucleoatac_signal.smooth.bw` files
Smoothed NucleoATAC signal with negative values removed for each developmental stage.

#### `.occ.bw` files
Nucleosome occupancy signal for each developmental stage. _Use this data for visualizing the position of nucleosomes._

#### `.occ.norm.bw` files
Nucleosome occupancy signal for each developmental stage, normalized for the number of reads per library.




## directory `NucATAC_eps/`
This directory contains visualizations generated by NucleoATAC. For more details, see the [NucleoATAC documentation](https://nucleoatac.readthedocs.io/en/latest/nucleoatac/).

#### `.Vmat.eps` files
"V-plot" for read insertion sites for each developmental stage.

#### `.nuc_dist.eps` files
Distribution of nucleosomes for each developmental stage.

#### `.occ_fit.eps` files
Modeled nucleosome distribution versus observed nucleosome distribution for each developmental stage.




## directory `Visualization_Files/`
This directory contains useful visualization files for IGV.

| filename | description |
|-------------:|:---------------|
| `OmniATAC_bothruns_q005_allpeaks.igv_new.bed` | Merged peaks across all stages. Used in ParhyaleATACExplorer scripts. |
| `OmniATAC_bothruns_q005_allpeaks.metrics.bed` | Merged peaks across all stages including useful summary statistics. |
| `Sall_peaktypes_distances_new.gff` | GFF file for all peaks, along with the assigned peak category (e.g. intronic, exonic), as well as distance to the nearest gene. |
| `Sall_peaktypes_nearestgene_new.gff` | GFF file for all peaks, along with the assigned peak category (e.g. intronic, exonic), as well as the name of the nearest gene. |
| `Sall_peaktypes_new.gff` | GFF file for all peaks, along with the assigned peak category (e.g. intronic, exonic). _Use this file for IGV visualization._ |



## directory `enrichment_specific_9acoresplit/`
This directory contains the enrichment statistics for transcription factors relative to a randomly-selected background (`match_specific/random_regions_smaller_mpbs.bed`), particularly for specific unique binding sites found in each of the 9 "acores" generated by Mfuzz.

___NOTE:___ Each "acore" assigned by Mfuzz is NOT identical to each Cluster referred to in the manuscript. The following table explains the relationship between each "acore" in these files and the final Clusters in the paper:

| acore number | Cluster number |
|-------------:|:---------------|
| acore6 | Cluster 1 |
| acore9 | Cluster 2 |
| acore5 | Cluster 3 |
| acore8 | Cluster 4 |
| acore3 | Cluster 5 |
| acore7 | Cluster 6 |
| acore2 | Cluster 7 |
| acore1 | Cluster 8 |
| acore4 | Cluster 9 |

### subdirectories `S#_hintatac_specific_acore#_unique`
Output data from rgt-motifanalysis enrichment. See the [Tool Usage](https://www.regulatory-genomics.org/motif-analysis/tool-usage/) page for details on output files.




## directory `enrichment_specific_small/`
This directory contains the enrichment statistics for transcription factors relative to a randomly-selected background (`match_specific/random_regions_smaller_mpbs.bed`).

### subdirectories `S#_hintatac_specific`
Output data from rgt-motifanalysis enrichment. See the [Tool Usage](https://www.regulatory-genomics.org/motif-analysis/tool-usage/) page for details on output files.



## directory `fraglengthcount_bothruns/`
This directory contains the fragment length distributions for each stage, merged between sequencing runs.

#### `.fraglengthcount.txt` files
Mapped fragment lengths for ATAC-Seq reads for each developmental stage.




## directory `match_specific/`
This directory contains the transcription factor footprints generated by RGT-HINT.

#### `_mpbs.bed` files
Original motif binding site prediction files generated by RGT-HINT.

#### `fixed_mpbs.bed` files
Modified bed files that remove trailing whitespace, which impedes bedtools analysis and IGV visualization. _Use this file for visualization purposes and enrichment analyses._

#### `random_regions_smaller_mpbs.bed` file
Randomly selected genomic regions with a size of 20 using bedtools random, used as background for enrichment analyses (see `enrichment_specific_small/` and `enrichment_specific_9acoresplit` directories for output results.)





## directory `match_specific_9acoresplit/`
This directory contains the transcription factor footprints generated by RGT-HINT, grouped by Mfuzz acores.

___NOTE:___ Each "acore" assigned by Mfuzz is NOT identical to each Cluster referred to in the manuscript. The following table explains the relationship between each "acore" in these files and the final Clusters in the paper:

| acore number | Cluster number |
|-------------:|:---------------|
| acore6 | Cluster 1 |
| acore9 | Cluster 2 |
| acore5 | Cluster 3 |
| acore8 | Cluster 4 |
| acore3 | Cluster 5 |
| acore7 | Cluster 6 |
| acore2 | Cluster 7 |
| acore1 | Cluster 8 |
| acore4 | Cluster 9 |

#### `acore#_wb_mbps.bed` files
bedtools intersect output for each acore vs. the mbps files.

#### `acore#_mbps.bed` files
Motif binding site predictions collected across all stages for peaks overlapping with a given acore.

#### `acore#_unique_mpbs.bed` files
Unique predicted binding sites for each acore. Duplicate footprints from across different stage-specific analyses are removed. _Use this file for visualization purposes and enrichment analyses._




## directory `q005_narrowpeak/`
This folder contains stage-specific peaks generated by Genrich.

#### `.narrowPeak` files
Genrich stage-specific peaks in [ENCODE .narrowPeak format](https://genome.ucsc.edu/FAQ/FAQformat.html#format12).
Data includes p-value for each peak and q-value merged across replicates by Genrich.

#### `.narrowPeak.gff` files
Genrich stage-specific peaks converted to [GFF format](https://genome.ucsc.edu/FAQ/FAQformat.html#format3) for viewing in IGV.




